### PR TITLE
Read custom cloud list from json string

### PR DESCRIPTION
**Pull request by JonCole**
This brings in the ability to specify the list of clouds from a json string, typically brought in through the grafana ini file.  A seperate PR on grafana/grafana will add the ini file settings.

### CODE REVIEW

File: azsettings/settings.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ type AzureSettings struct {
 
 	// This field determines which plugins will receive the settings via plugin context
 	ForwardSettingsPlugins []string
+
+	customClouds []*AzureCloudSettings
```
Comment (JonCole) [inline azsettings/settings.go:26]: I realized we may need to export this customClouds field after all - I was trying to keep it private to the package to push callers to use the functions, but inside grafana/grafana, we need to write this list back out as a json string when we push settings over to plugins via env variables.  Thoughts?

Comment (aangelisc) [reply azsettings/settings.go]: I think this should be fine to remain private for now. In `grafana/grafana` we can still call the `Clouds` or `CustomClouds` functions to retrieve the values.

File: azsettings/settings.go [hunk 2]
```diff
@@ -60,6 +62,11 @@ func ReadFromContext(ctx context.Context) (*AzureSettings, bool) {
 		settings.AzureAuthEnabled = true
 	}
 
+	if customCloudsJSON := cfg.Get(AzureCustomCloudsConfig); customCloudsJSON != "" {
+		// this method will parse the JSON and set the custom cloud list in one go
+		_ = settings.SetCustomClouds(customCloudsJSON)
```
Comment (aangelisc) [inline azsettings/settings.go:67]: I suggest adding a log for this error instead of ignoring it.

**Review by aangelisc — COMMENTED** [review]
This LGTM, just one nit in the comments. Also, what's triggered the dependency updates?

File: azsettings/settings.go (hunk 1 above)
Comment (JonCole) [reply azsettings/settings.go]: `Clouds()` and `CustomClouds()` return `AzureCloudInfo` struct, which does not have all the fields that are on `AzureCloudSettings` struct.  `grafana/grafana` needs the AzureCloudSettings to get the additional information.  I think we might eventually want to get rid of the `AzureCloudInfo` and just return the full set of fields in one struct and then any caller can use the same exact set of methods for all use cases.  If a caller doesn't need the additional fields, they can ingore them.  Thoughts?  For now, I was just leaning toward exporting `customClouds`.  Thoughts?

**Comment by JonCole** [comment]
> This LGTM, just one nit in the comments. Also, what's triggered the dependency updates?

TBH - I didn't notice exactly what is bringing in the dependency updates - I will try to clean those up if I can and still get it to build.

**Comment by JonCole** [comment]
> > This LGTM, just one nit in the comments. Also, what's triggered the dependency updates?
> 
> TBH - I didn't notice exactly what is bringing in the dependency updates - I will try to clean those up if I can and still get it to build.

Looks like dependabot did the dependency updates, not me - see commit: https://github.com/grafana/grafana-azure-sdk-go/pull/141/commits

File: azsettings/settings.go (hunk 1 above)
Comment (aangelisc) [reply azsettings/settings.go]: I'm fine with exporting the field for now, there's nothing that would be considered a secret in there afaict

File: azsettings/cloud_settings.go [hunk 3]
```diff
@@ -68,15 +69,28 @@ func (*AzureSettings) GetCloud(cloudName string) (*AzureCloudSettings, error) {
...
-	clouds := getCustomClouds()
-	return mapCloudInfo(clouds)
+func (settings *AzureSettings) CustomClouds() []AzureCloudInfo {
+	return mapCloudInfo(settings.CustomCloudList)
+}
+
+// Parses the JSON list of custom clouds passed in, then stores the list on the instance
+func (settings *AzureSettings) SetCustomClouds(customCloudsJSON string) error {
+	var customClouds []*AzureCloudSettings
+	if err := json.Unmarshal([]byte(customCloudsJSON), &customClouds); err != nil {
```
[the remark this answers is not in the document]
Comment (JonCole) [reply azsettings/cloud_settings.go:85]: Sure, that type of change would be fine. I will try to push that today but if I can't, feel free to add that when you go to test it out

**Review by aangelisc — APPROVED** [review]

**Pull request merged**